massa-protocol-worker/src/protocol_worker.rs

```diff
@@ -1234,50 +1013,231 @@ impl ProtocolWorker {
+                                        // check thread
+                                        if op.thread != info.header.content.slot.thread {
+                                            massa_trace!("protocol.protocol_worker.process_block.err_op_thread",
+                                                 {"block_id":block_id, "op": op});
+                                            should_be_banned = true;
+                                            break;
+                                        }
+
+                                        // Check for duplicates
+                                        if !received_ids.insert(op.id) {
+                                            should_be_banned = true;
+                                            break;
+                                        }
+                                    }
+
+                                    // Check the ids are exactly what is expected.
+                                    if !(wanted_operation_ids == &received_ids) {
+                                        should_be_banned = true;
+                                    }
+
+                                    if should_be_banned {
+                                        let _ = self.ban_node(&from_node_id).await;
+                                        return Ok(());
+                                    }
+
+                                    // Re-constitute block.
+                                    let endorsement_ids = info.endorsements.clone();
+                                    let block = Block {
+                                        header: info.header.clone(),
+                                        operations: info.operations.clone().unwrap().clone(),
+                                    };
+                                    let content_serializer = BlockSerializer::new();
+                                    let mut content_serialized = Vec::new();
+                                    content_serializer
+                                        .serialize(&block, &mut content_serialized)
+                                        .unwrap();
+                                    #[cfg(feature = "sandbox")]
+                                    let thread_count = *THREAD_COUNT;
+                                    #[cfg(not(feature = "sandbox"))]
+                                    let thread_count = THREAD_COUNT;
+                                    let wrapped: WrappedBlock = Wrapped {
+                                        signature: info.header.signature,
+                                        creator_public_key: info.header.creator_public_key,
+                                        creator_address: info.header.creator_address,
+                                        thread: info
+                                            .header
+                                            .creator_address
+                                            .get_thread(thread_count),
```

Comment:
```suggestion
                                        thread: info
                                            .header
                                            .thread;
```


Important reminder:

* an address belongs to the thread index defined by taking the first 5 bits of the enclosed hash
* an operation belongs to the thread of its sender's address
* the thread to which a block belongs is not defined by its hash, nor by the thread of the address of its creator, but by an explicit `slot` property inside its header
* a block belonging to the thread T can only include operations that belong to thread T
* any address can produce blocks that belong to any thread
* an operation belonging to thread T can spend and receive parallel coins to and from addresses belonging to any thread
* an operation belonging to thread T can make an address belonging to any thread receive sequential coins
* an operation belonging to thread T can make addresses belonging to thread T spend sequential coins.
* an operation that belongs to the thread T cannot cause sequential coin spending from addresses that do not belong to thread T.

Reply:
@gterzian